### ISSUE TITLE
labwc on fractional scales incorrectly (blurry) renders view

### ISSUE THREAD

**Issue by dmitry-azaraev** [labwc/labwc#1527]
When using fractional scale, e.g. 1.25 or 1.5 at fist looks everything look nice...

However, if set window in right position (just by slowly moving), then it is possible to notice some bad effects:

- SSD borders become one pixel widen than expected, top border becomes even worse - it's height increases only between two buttons (which is not rounded in my case);
- SSD button layout little shifts (maximize button may jump a little);
- At some really "unlucky" window positions text become more blurry: this can be observed at labwc's root menu, but also can be observed in application(s), e.g. chromium in wayland mode may blur whole view, this behave like whole application view get rendered at "fractional" position, which never must happens by any case by any means. Note, what when app content blurred - SSD borders may be clean, so this is kind of different thing. Is separately may applied to both axis.
- Same font blurrness can be noticed at SSD title

Also, nested labwc is just blurry, but it free from effects described above.

I'm tried to digg myself, but looks like I'm looking in wrong places. Probably problem rises directly from wlroots? But I'm have no idea what to do. Maybe some one can give me some pointers about?

**Comment by dmitry-azaraev** [comment]
I'm should sorry because I'm trick even myself, however, described behavior in issue still applicable.

I'm found what I'm did not actually understand HiDPI support in labwc. :)

I'm tried to use scale=2 and even scale=4 and theme'd SSD borders is *scaled* (while I'm thought what they should not). E.g. border.width=1 becomes clearly not a 1 pixel, but 2 and 4 pixels. At scale=2, top-left corner (with corner radius = 0) is overlaps with border at 1 pixel, and get mixed. I'm did not configured screenshot grabbing yet so can't add image, sorry. :)

For easy testing this just define embedded scheme in `rc.xml`:

```
<labwc_config>
  <theme>
    <name></name>
    <cornerRadius>0</cornerRadius>
  </theme>
</labwc_config>
```

And put `themerc-override`:
```
border.width: 1
window.active.border.color: #ffffff
window.active.title.bg: flat solid
window.active.title.bg.color: #33393B
```

At same settings incorrect border drawing (like I'm described initial post) is also visible at scale=1.25.

Below I'm put some statements, it is basically just mine expectations, correct me where I'm wrong.

1. Wayland-way to deal with HiDPI is provide to clients surfaces which will be rendered to output device in 1-to-1 pixel. E.g. scale property is actually only property/hint to client which might use it to scale content, while actual scaling during composition (normally) should never occurs.
2. As implication from (1) is what client's surfaces/views should be pixel-perfect positioned on final output device. (There is no point to perform pixel-perfect rendering and then just shift it by 0.5px making everything blurring)
3. Themes (i mean SSD mainly) generally should be rendered in pixel-perfect manner. E.g. border should wrap client view without any gaps. When on non-fractional scale is natural, but imagine border which is 1.5x width: with subpixel rendering and/or antialiasing is easy to draw lines with such width, but resulting width will be 2 device pixels, and inside this 2 device pixels line can be "bound" to one of edges.

At this moment, I feel what labwc (or wlroots) doing (3) wrong even on non-fractional scale, however, I'm still did not understand source of this (both labwc and wlroots is new to me). However, to properly "fix" this - labwc should state how it would deal with theme. If theme values is scaled: then sizes should be able be fractional, like css pixels can be. If theme is un-scaled, then it is should be possible make theme for target scale.

As for app view positioning - is probably not labwc issue, because I'm see it only with chromium and it might be chromium's issue as well (probably similar issue): https://issues.chromium.org/issues/40934705 . However, for me it reacts on window position, while kitty and foot - rendered perfectly no matter how I'm moving window.

> We do not set subpixel modes for outputs

Yes, subpixel mode should be set, even while defaults works okay. Default/unknown mode is backend-dependent, and this is bad. I'm not sure if it is tied, but if I'm use WLR_RENDERER=vulkan - fonts have some purple edges. But on mine amdgpu integrated graphics vulkan is looks like buggy, so it is probably not relevant. But I'm not sure where is right place to set subpixel mode in labwc, actually tried yesterday in some random place, but did not had chance to complete test yet.

> We do not tell wlroots what filter to use to scale scene graph items

Should be easy? But same problem for me. :)

PS: It is exist way to dump scene with some tool?

PPS: I'm using `rofi` (Xwayland), and fonts get bit blurry in 1.25x scale. And generally lot of software is not ready to scaling, so I'm even not sure, what I'm really want to use 1.25 scale. :(

**Comment by Consolatis** [comment]
> PS: It is exist way to dump scene with some tool?

Yes, the undocumented `Debug` action will dump the scene tree into the log.

**Comment by dmitry-azaraev** [comment]
Thanks!

`Debug` action works well (i'm using latest version from git). However, I'm suggest flush stdout, otherwise end of tree is not visible.

```
diff --git a/src/debug.c b/src/debug.c
index af722f3..c8e0f3e 100644
--- a/src/debug.c
+++ b/src/debug.c
@@ -232,6 +232,7 @@ debug_dump_scene(struct server *server)
 	printf("\n");
 	dump_tree(server, &server->scene->tree.node, 0, 0, 0);
 	printf("\n");
+	fflush(stdout);
 
 	/*
 	 * Reset last_view so we don't access a
```

As for mine issue - I'm spent a little around, and it is now way more clear why it is did not work, as expected by me.

That's because labwc just setup output scale and wlroots scale everything by this. This means, what scene coordinate space is also scaled (down), and by so - it looses precision. I'm not sure what is correct way to deal with from wlroots perspective (can we "scale up" scene space?), but even so, - there is generally fundamental flaw in design in my taste: coordinate space should be able represent any pixel on screen exactly.

And even worse: to support dynamic scale changes to keeping window positions - they should be "overscaled" (e.g. in situation when changing from 4x scale to 1x scale, scene space still should express correct positions for new scale), however, there is possible just by adjust them once.

I spent some time with 1.0x, and found what I'm absolutely want to have normal scaling (e.g. 1.25x) in first place... something must be changed. It is exist some vision how exactly "fractional scale" should work? I know about fractional-scale-v1 protocol, but, it is not what all applications rely, and I'm not sure what it is enough. Chromium for example have just implementation stub for this protocol which do nothing at this moment.

PS: I feel what in current implementation - SVG theme buttons are rasterized and only then up-scaled. :tongue:

**Comment by ahesford** [comment]
It's not clear to me that representing physical pixels instead of logical (scaled) pixels in the fractional-scaling protocol or wlroots would be an improvement over the current state of affairs. Yes, it would be possible to address each physical pixel directly. However, it would also involve extra bookkeeping. With logical pixels, if you move (or extend) a window across two displays with different scales, the size on both displays is consistent because they use the same logical pixels as a reference unit. With physical pixels as the ruler, a window will change apparent size as it moves to another display. This is annoying enough to cope with, but consider a worse case, where the window resides across two outputs (or an output scene is mirrored on a display with lower resolution). Suddenly, the window can't be represented as a single rectangle, because the units are disparate on both displays. A lot of simple math becomes a lot harder.

Pixel-perfect output, if necessary, can be achieved with unit scaling and DPI-aware clients. In most cases, the goal is to make pixel-perfect positioning at arbitrary scales irrelevant by ever increasing display resolution.

**Comment by Consolatis** [comment]
> PS: I feel what in current implementation - SVG theme buttons are rasterized and only then up-scaled. 👅

Yes, that is currently an issue. To make that work properly labwc has to dynamically render the SVG icons whenever the scale of one of the scene_buffer users of the SVG `lab_data_buffer`s changes. It should also provide an LRU cache of the SVG `lab_data_buffer`s for the different scales so we don't end up rendering the icons over and over again. Similar to the autoscaling title and menu buffers but its a bit complicated here because we have multiple `wlr_scene_buffer` wrappers around them. To sum up: we need some generic changes in labwc how we handle theme buffers before we can sensibly make the SVG icons render on-demand at the correct scale.

**Comment by dmitry-azaraev** [comment]
@ahesford there is not necessary should be physical pixels, coordinate space just should have enough precision to hit in pixel under different scales. For example it can be 1/64 of logical pixel. But I'm doesn't see what it matters at all - I'm can't find any way to change this via wlroots API.

> Pixel-perfect output, if necessary, can be achieved with unit scaling and DPI-aware clients. In most cases, the goal is to make pixel-perfect positioning at arbitrary scales irrelevant by ever increasing display resolution.

Before clients, labwc draw some things by self: SSD, title, menu - all of this get's blurry just depending on position. So mine worries mostly about this. As for clients, they gets eventually fixed, but right now chromium view get's blurry on some window positions, in font rendering. Firefox have similar issue too, but it hard to see on fonts, but it is very easy to see on just horizontal line and moving window vertically. I'm doesn't want debug chromium before I'm be sure what compositor doing own job well.

**Comment by dmitry-azaraev** [comment]
@Consolatis thanks, - `hidpi-scaling.html` is good and helpful page, I'm guess I'm installed kanshi exactly from this suggestion. I'm did not go into override environment variables yet, just because I'm stuck with labwc theme + chromium, and no other application require tweaking yet. I'm only stuck with good RDP client, but wlfreerdp looks like works, at least with 1.0 scale, and even then it have some scaling burden (i guess at protocol level).

PS: I'm bind ToggleKeybinds to W+F6 and it works perfect, except what when toggle triggered - looks like W button remains in pressed state for client (so need press it again). Also when keybinds are disabled/inhibited - mouse binds works anyway - e.g. ALT+mouse click handled by labwc, but it should be handled by client application in this case (and this is pretty common combo, - part of ALT+TAB switching).

**Comment by Consolatis** [comment]
> I'm bind ToggleKeybinds to W+F6 and it works perfect, except what when toggle triggered - looks like W button remains in pressed state for client (so need press it again)

This sounds very much like a bug within our keybind handling. It is notoriously hard to get right. Does this only happen on `ToggleKeybinds`? In any case, feel free to open an issue about it and also mention what version of labwc you are experiencing this with.

> Also when keybinds are disabled/inhibited - mouse binds works anyway - e.g. ALT+mouse click handled by labwc, but it should be handled by client application in this case (and this is pretty common combo, - part of ALT+TAB switching).

This is a good point, mind to open an issue about it?

**Comment by dmitry-azaraev** [comment]
@Consolatis yep, i will open issues.

**Comment by dmitry-azaraev** [comment]
@ahesford I'm understand what fractional scales are not perfect, but they are very common by itself. QHD becomes more popular across world, phones usually have something strangle like 2.75 or so. Chromium/Blink for example uses fixed-point math with 1/64 during document layouting. But subpixel positioning doesn't solve problem alone: during rendering browser(s) apply lot of effort to snap layout to whole pixel boundaries. wlroots here did not make good job currently: it just multiplies and rounding every component of box, but most likely it should pick enclosing box (e.g. whole pixels, not "random" rounding).

I'm personally prefer logical pixels, however, Windows, for example, uses raw pixels instead in window manipulation functions for dpi-aware applications.

> I believe the only sensible choices are: 1) defining logical pixels as they are now, so that there is a consistent frame across all displays; or 2) just use physical pixels and expose the arithmetic complexity to clients and compositors.

1) I'm not sure what clearly understood what you mean here. Use 1/120? Technically just now logical pixels is not defined by wlroots, as it simply pick scene geometry and multiply it to output scale, and by so, right now scene coordinates are fixed at 96dpi. As for 1/120, I'm think what scene not necessary should follow exactly same multiplier, as it is used only(?) for express scale to clients via protocol, while wlroot's compositor scene is something private to him (e.g. 1/64, 1/128 may work).

2) Yep... logical pixels feels much more better here :)

In both cases, looks like it is wlroots issue in first place. Hum.

**Comment by dmitry-azaraev** [comment]
I'm got a small progress:

1. chromium no more randomly blurs, it draws same at any position.
2. firefox stops to blur horizontal line.

In wlroot scene I'm tweak rendering a bit:

```diff
diff --git a/types/scene/wlr_scene.c b/types/scene/wlr_scene.c
index f81a8613..b983be8f 100644
--- a/types/scene/wlr_scene.c
+++ b/types/scene/wlr_scene.c
@@ -891,14 +891,14 @@ static void scene_node_get_size(struct wlr_scene_node *node,
 }
 
 static int scale_length(int length, int offset, float scale) {
-	return round((offset + length) * scale) - round(offset * scale);
+	return ceil(length * scale);
 }
 
 static void scale_box(struct wlr_box *box, float scale) {
    box->width = scale_length(box->width, box->x, scale);
    box->height = scale_length(box->height, box->y, scale);
-	box->x = round(box->x * scale);
-	box->y = round(box->y * scale);
+	box->x = floor(box->x * scale);
+	box->y = floor(box->y * scale);
 }
 
 void wlr_scene_node_set_enabled(struct wlr_scene_node *node, bool enabled) {
```

So every time it render scene entry - every box always snapped by determined algorithm, and it is represents enclosing box. scale_length (e.g. `round((offset + length) * scale) - round(offset * scale)`) - is a just hacky workaround for something. I'm did not know wlroots good enough, probably somewhere exist implicit assumptions about, we still want to have surface size equal to displayed size.

Also, from wlroots code I'm found what actually using logical pixels probably is a only way, because it might render same scene to multiple outputs with different scale. Using device-pixel-scale as base is nice, and probably easier to get work, but it then ends what second output need to be somehow scaled again.

However, it is opens and demonstrates already existing holes:
1. `foot` renders strangely: it keeps lot of cursor artifacts, which disappers after moving window :) I guess it is because it uses subsurface until some point, but there is still just wrong positioning. In contrast `kitty` doesn't uses subsurface and everything fine. Also, if properly align window, then foot also did not have artifacts while typing.
2. it is shows what there is not enough precision in scene graph.
3. fonts in labwc title now constantly blurry near center, and less blurry to edges (as title becomes longer and longer)
4. At some position 1-pixel SSD border becomes 2-pixel border (this happend and without this change) - there is because border width 1.25 is already 2 whole pixels, and second pixel overlaps with client area. In some cases client area just overlap border, and in other - not.

There is exactly case where pixel snapping is required (e.g. client area should be snapped), and this part missing now in official version. However, the problem what and SSD and client area requires snapping, and in current coordinate space it is can't be expressed. And if it will be possible to express this - then it will need to rethink theme, how treat line widths, adjust them according scale, and generate proper scene graph. I think, something similar can be achieved by extending scene graph instead, adding notion of snaps. In that case, probably, current scene coordinates might work as is.

Another workaround which might somehow work - is always align windows to some positions, which is compatible with current scale. :) However, I'm not sure if I'm ever might live if windows will be moved in steps of ~5 pixels. Very not universal... Huh. Sorry for wall of text.

**Comment by dmitry-azaraev** [comment]
With conjuction of patch above, when title rect get's stable physical size - blurry text in labwc's title and menu can be easily fixed:

```
diff --git a/src/buffer.c b/src/buffer.c
index 0921acd..1bb59c2 100644
--- a/src/buffer.c
+++ b/src/buffer.c
@@ -91,8 +91,8 @@ buffer_create_cairo(uint32_t width, uint32_t height, float scale,
 	struct lab_data_buffer *buffer = znew(*buffer);
 	buffer->unscaled_width = width;
 	buffer->unscaled_height = height;
-	width *= scale;
-	height *= scale;
+	width = ceil(width * scale);
+	height = ceil(height * scale);
 
 	/* Allocate the buffer with the scaled size */
 	wlr_buffer_init(&buffer->base, &data_buffer_impl, width, height);
```

Now it looks nice and did not depends from position also. Yup! It is one pixel hunt, which completely crisp image.

Using  WLR_SCALE_FILTER_NEAREST also possible, but without correct geometry it will "remove" pixel line from center, crisping at least one letter, so, looks inaccurate. I'm did not know if using nearest is more performant, but I guess it's renderer (wlroots) job to not use bilinear filter if it is not necessary (and I'm doesn't see any evidence of this in code).

**Comment by dmitry-azaraev** [comment]
So, now looks good, and i'm ended here (somewhy PNG is upscaled after passing via clipboard/or it is chrome upscale, actually it looks clearer):
![image](https://github.com/labwc/labwc/assets/14053068/6a863071-632b-4e5b-9fb5-2348991a1f4a)
I'm resized window specially to show oversized 2px borders, instead of 1px, and overlapped corners.

Additionally, I'm edit `debug.c` to print W and H of nodes (may require cleaning): [debug.PATCH](https://github.com/labwc/labwc/files/14307069/debug.PATCH)

Now, I'm lost, need idea what is possible to do with SSD rendering... Just increasing scene precision in wlroots will not solve issue alone anyway, even if I'm feel it right way. wlroot render boxes and scale them idependently. Probably in-tree-dependent scaling might help, but this requires some refactoring, and I'm did not want end with own fork of wlroots. :thinking:

**Comment by dmitry-azaraev** [comment]
I'm disappointed and give up...

I'm doesn't see any sense to dig into this, just because every related thing which I'm touch - is broken. There is feel same as 20 years ago, with exception, what nowadays linux kernel can handle any hardware well out of box. But UI still not exist at all, UX - is even worse, it is almost no one even heard about (well, labwc authors heard about). Sorry, but I'm will rant now, so did not read if don't want read ranting.

In addition of all this burden described above, added mainly by wlroots... And, where I'm come? I'm spent enough time to see wlroots code, and found what it is  just yet another bad library with bad design, bad coding and bad decisions. I'm out. I'm doesn't know why peoples think what is good, but it is coded badly almost everywhere. Even language choose is questionable. C?! Yep, it is smaller, but it is not faster, and most likely it is burden simple things, into unmanageable macroses or other thing. Compositor is just compositor, it is perform work, but it is medium-loaded by design. It have no requirements like kernel. It can be written in C# or Java and work even better, than it works now, because there is absolutely no matter what you code, almost all it's code called very rarely. It completely missing ANY comments. It missing documentation as well. Want to see GOOD compositor code? Look chromium, it had good compositor which works on every platform and can handle layouts which way more complexier than wlroots do with his "scene". Chromiums compositor works in every desktop browser, and serve all window/compositor purposes on ChromeOS. And it have at least comments, and when you open file, it usually defines purpose. Not always, but generally it is true. So, no - wlroots - is BAD library in every piece. There is no need to argue about, there is my just personal opinion.

Patches added above in this issue solve window-position issues, making surface mapping stable, but not solve window-size issues. It still just can map texture to one pixel wider or smaller area, and it will be completely blur, - you might hit what one menu is blurry, other is not. How to avoid this?! Huh, is simple. There is needed just stop doing stupid things, which wlroots doing.

QT5: Using doublecmd at 1.0 scale menus already broken - shortcuts rendered as glyph underline sometimes becomes 2 pixel width/height. At 1.25 scale this things actually looks better, but it is broken rendering in first place before scaling come. Great?! Guys, you did not snap to pixels every menu entry?! Is easy and bullet-proof. It is bad from UX perspective when glyph looks differently just because it is rendered somewhy at different subpixel. There is why exist flow, block elements, etc. There is nonsense, and there is why, I'm dislike QT. GTK have no such issue, and never had (I guess).

labwc's "scene" - is good idea, but it should be layouting engine, but it is not - it is just put boxes from input and put on random output pixels. Damn. Why anyone ever needed library for this?!?! Desktop Window Manager/Compositor surely should handle all of this stuff, but it MUST be intelligent to not doing not needed work. In fact - I'm reversed thoughts about wlroots after inspecting it's code - I'm doesn't want to see this library in my installation completely. wlroots will do heavy work for nothing, like bilinear mapping every texture, while it should not do it in every desktop setup, unless some corner conditions. In normal conditions it should not. But, it is not only do it - it is default mode.

**HOWEVER**

Mine actual UX problem is font scaling under wayland with 1.25x. Somewhy it is very different from doing same but with just increasing font size (and labwc's font drawing looks correct in both cases). With 1.25x output scaling glyphs just becomes significantly wider than they should, and it is very annoying... Just imagine what your terminal font wider than taller. It is not in practice, but font rendering broken in every application. It is not specific to wlroots I guess, but it is looks really bad. 1.25x scale should not make fonts more wider nor stronger.

I'm will try setup by just using 1.0 scale, and increasing font sizes (this will solve any positioning issues in wlroots, and by so, labwc will work correctly with current theming). However, chromium is problematical case in this case - with wayland backend it can't be scaled up without native mechanism (and per-page scale factor is not what anyone really want to use, it is useless option at all). And generally I'm dislike this idea because I'm want to see correct rendering in any applications, css have physical units, and they should work. However... screenshot ting with 1.0 scale works correct, while at 1.25 like on image in previous post - incorrectly scaled too for no reason.

Many software just did not ready to non 1.0x scale. But how software will, if compositors core can't agree with client which width of window it should present?!?! No. I'm definitely quit from this joke. wlroots is history.

Any labwc minor things mentioned here should be addressed separately, but I'm doesn't see any sense to do this unless wlroots be able to produce correct output in first place. Instead it doing everything what conforms "foot", which render incorrectly in first place, and then wlroots extend damage regions for fractional scale, which again stupid thing (because there is should be just universal equation instead of special casing). Nothing will change in 0.18, i guess, probably will be bit better, but it almost same bad as 0.17 (but it feels bit differently, but random rounding still in here). I'm can't accept what during PR/CL used third-party applications, which adapt to wlroots itself, instead doing math tests, how it should work. So, in addition to absense of any kind of tests, they are test changes against software which already had been created for. Huh. It is expected, it is so "swayi" (youngy).

This doesn't means what I'm would stop use labwc, but my plan now changed: I'm wish to trying writing compositor itself (but time is limited, this doesn't mean what I'm will do it right now).

I'm really very appreciate labwc project, I'm like project owners live feedback in issues (even which is not mine), which often ends in discussion not related to scope, - is looks like warm and good community. And it configurable like needed.

However, labwc scope far from ideal, real peoples needs "normal" taskbar with window preview, with window pinning, normal window switchers with preview, and this all out of labwc scope. I'm can live without this, and even more, it can be implemented, but having underlying technology (wlroots) which can't position surface properly - there is little more than I'm can handle, and want to see. I'm already seen enough performance fuck-ups with Windows's dwm.exe - so, no. I'm out.

Theming - is another uncovered place. I'm, for example, modify button width for self, because it is just constant in code, and I'm need square buttons, and title height is higher than constant. I'm understand what there is always exist more serious things, but theming in current labwc state is not existent, and "openbox specification" is give nothing to project - no theme compatible with it actually in current stance, because labwc did not implement more than 70-90% of defined options. But, theming is kind of thing which I'm not rant: is eventually evolved and it is predictable and relatively easy, or can be predictable contributed. But, I'm will not even try contribute in this scope, because at current wlroots state, there is no way to implement themes correctly, because there is impossible to align outer and inner boxes in 1 pixel, and there is no matter if you will create single node/surface or multiple like now.

This issue was about eventually blurry application views, **but problem is already tracked down**, and it is all about wlroots, not labwc (or very minor for labwc), so I'm closing issue.

**Comment by Consolatis** [comment]
Thanks for sharing your conclusion. I will only reply to 2 things here:

> However, labwc scope far from ideal, real peoples needs "normal" taskbar with window preview, with window pinning, normal window switchers with preview, and this all out of labwc scope. 

Yes, these things are indeed out of the labwc scope. A "normal" taskbar with window pinning would be easy to implement by a third party though (and is one of the intended goals of labwc: users are free to choose what kind of taskbar they want to use and if they want to use one at all). https://github.com/LBCrion/sfwbar is a good project with a very response dev and implementing pinning should be relatively trivial. In regards to the previews, those could be done once the [screencapture protocol](https://gitlab.freedesktop.org/wayland/wayland-protocols/-/merge_requests/124) gets accepted as it allows captures of toplevels additionally to full outputs. But I do get your point about the restricted feature set.

The other point is just a suggestion: creating an issue at wlroots which describes the generic scaling issue and suggest a way to solve it long term. That obviously doesn't mean that you have to invest your time into coding it up but it might be worthwhile to exchange some ideas so that in the long term the issue could be tackled properly. You seem to be pretty knowledgeable in the area.

**Comment by dmitry-azaraev** [comment]
@Consolatis I'm probably try again sfwbar but it is in AUR on Arch Linux, means what I'm should care about personally (not a big deal). However, I guess it did not expose previews anyway. Oh, should appreciate for ability to build labwc with custom/embedded wlroots easily. :) I'm currently stick with waybar which somehow work (not without issues), but I'm actually with this perfect-scaling stuff did not even care too much about. Damn, I'm even did not setup background. Surely, I'm really like XFCE but it is not ready (for wayland), and I'm willing something different. I'm had custom setup about 8 years ago, but it was at X11 of course with tint2 as taskbar I guess (I'm did not remember exactly). Also I'm know what Cinnamon ~8 years ago was only one distribution who properly handled Alt+Shift keyboard layout switching, because every other fail pass this modifiers within terminal when you trying press shortcut in Midnight Commander for example. I'm trying avoid all of this "wide" distributions stuff, because many things changed (not so much in XFCE, and there is bad thing - they should move faster). But I'm also doesn't want see windows where I'm did not know how to close window, it is why I'm like server decorations and first place (i mean, modern past GTK3 apps). I'm also like client decorations, like chromium was integrated in Windows decorations past years - almost seamlessly and very naturally, starting from Windows 7. But they changed theme in 2023, so now it is looks way more elegant, but it is still option, but it is mess. There is two modern "UX" now: (1) you have oversized close button and (2) you can't find close button, and see bucket of random buttons. Nothing of this I'm like. But this is only visual. But from UX perspective, I'm tried recently "raw" gnome on debian. Nice what it install FAST, not so slow like 15 years ago, but it is unusable for me. I'm liked gnome 2 (i know about forks, but they in practice doesn't matter a lot, as anyway I'm want use modern software). I'm liked Unity and this stuff with external menu in top bar. KDE i'm dislike, it really looks familiar to windows (which I'm use every day), but I'm doesn't like QT/KDE just religiously from early 2000 (e.g. without good reason), but in modern KDE I'm can get lost just in window/compositor menu. It have too many options. :) Can I be little bit religious and avoid KDE stuff? Also, I'm did not want get same Windows-like workflow, where I'm spawn lot of consoles and only when there is too much / when I'm remember - I'm close them. Nor labwc/nor KDE/nor gnome did not solve this completely, but instead often adds new challenges instead of top of existing. I'm tried tiling managers, but there is completely not mine thing (i like idea, but I have poor sight, I'm can concentrate on center of display, and leaving sides for not important stuff, but I'm absolutely can't split and lose concentration, is nor fancy, nor productive).

BTW: Window placement in labwc is something what can be extended. Now, is have good options, but I'm prefer just center, not because it is best, just because it is less disturbing than other modes. Window placement (for me personally) thing already invented in Windows - whenever default placement is intended it derived from current window, and shifted a bit. I'm actually have no idea, because it is anyway often ends moving, but often you don't care, and it smart enough. The key what it did not trying "tile" them, which in labwc mode nice, but without clear understanding what happens is no matter (without good taskbar) - how you not place them, they will overlap. Add here RDP machine, and there is total mess. :) So, I'm just stick with center - at least new console appears over my eyes, and I'm doesn't need to move head.

> But I do get your point about the restricted feature set.
I'm did not come randomly, I'm generally carefully read labwc scope before and goals and non-goals at least two times. :)
For theme which I'm want want (from screenshot you should see, it is pretty simple, except what borders on screenshot is white for demonstrating issue), and I'm should say what I'm agreed on hard stance, - protocols should be involved instead of implementing custom things. However, in practical sense - we want having good task switcher and etc right now, no matter how it is implemented. And, there is possible even maintain (in fork) in case, if some additions never be merged. I'm already say, but repeat again, I'm just mention, what I'm want more, and plan to make some things myself. But right now I'm did not see sense, just because they did not work anyway. I'm actually pretty impressed with labwc. It is kind of project which just work in my experience.

Mine main disappointing is not wlroot, coding style or something. I'm disappointed what wayland was started relatively recently and scaling stuff had initially has not been covered, and I'm doesn't see any reason for what, because when wayland starts to grow - there is was lot of devices with nor 1 nor 2 scale. I'm disappointing what Linux Desktop is not a thing, at least not out of box. There is exist plenty of good distributions, which looks good, but there is not means what they are really handy in use. I'm disappointed what Ubuntu, which I'm liked before - moved to snaps, which definitely slower, and it is kinda of not necessary stuff. I'm understand what it is easier to build app for some library environment, but proper Linux distribution have package repository with native binaries which match one to other. Even it can add some "dll-hell", but snap is not cost-free, and very not cost free. From my understanding best thing which we get from snap (ouch from flatpak actually?) is xdg-desktop-portal. I'm using ArchLinux just because I'm use it before (about 8 years ago), now just come time to get back, to setup system, which has higher requirements (aside from UI), and generally I'm always liked Linux, and administering Windows and Linux at beginning of 2000.

Still, I'm disappointed, what nothing changed. New protocol (Wayland), support for new protocol is added and it is live! It is great. But nothing changed. :) 

Also, new protocol is great, but how you thing browser during webrtc sessions should filter keyboard noise without global keyboard hooks? Chrome doing this on windows, windows allow do it. Well, I mean it is not question, I mean, wayland on mine closer looks generally looks right, some errors already done many years ago, but I'm like architecture. However, some questions are uncovered, which are tried to fixed with additional protocols. Sometimes they rejected by security reason, but example with webrtc key noise filtering is also security reason. :)

PS: I'm need cooldown before posting issue to wlroots (and need to register at freedesktop :) ). I'm not sure what I'm want to register and post issue and participate. Probably I'm will try do it later, but right now I'm did not want do it. Also, I feel what 0.18 will be bit better, it will be still buggy as now, but visually should be better I guess (wltiny with manually adjusted scale  from master at least do it). So, it is anyway exist big gap between wlroots-main and labwc-main, so I'm definitely would like to wait 0.18, and the think. However, question actual in 0.18, I'm sure. But, I'm not ready express it anyway.
